File: LeetCode/leetcode-30day-challenge/week3/C++/ValidParenthesisString.cpp
#include <bits/stdc++.h>
using namespace std;

//Type: Stacks, strings

//Approach: Implementation
//Complexity: O(N) time , O(N) space complexity

class Solution
{
public:
    bool checkValidString(string s)
    {
        stack<int> parStack;
        stack<int> starStack;

        for (int i = 0; i < s.length(); i++)
        {
            if (s[i] == '(')
            {
                parStack.push(i);
            }
            else if (s[i] == '*')
            {
                starStack.push(i);
            }
            else
            {
                if (!parStack.empty())
                {
                    parStack.pop();
                }
                else if (!starStack.empty())
                {
                    starStack.pop();
                }
                else
                {
                    return false;
                }
            }
        }

        while (!parStack.empty())
        {
            if (starStack.empty())
            {
                return false;
            }
            else if (parStack.top() < starStack.top())
            {
                parStack.pop();
                starStack.pop();
            }
            else
            {
                return false;
            }
        }
        return true;
    }
};

File: LeetCode/leetcode-30day-challenge/week3/Java/ValidParenthesisString.java
import java.util.Stack;

class Solution {
    public boolean checkValidString(String s) {
        Stack<Integer> stack = new Stack<Integer>();
        Stack<Integer> starStack = new Stack<Integer>();

        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '(')
                stack.push(i);
            else if (s.charAt(i) == '*') {
                starStack.push(i);
            } else {
                if (!stack.isEmpty()) {
                    stack.pop();
                } else if (!starStack.isEmpty()) {
                    starStack.pop();
                } else
                    return false;
            }
        }

        while (!stack.isEmpty()) {
            if (starStack.isEmpty()) {
                return false;
            } else if (stack.peek() < starStack.peek()) {
                stack.pop();
                starStack.pop();
            } else
                return false;
        }
        return true;
    }
}